Update tests to permit any order of output lines by sorting the editorconfig output

cxw42 opened the pull request:
@xuhdev @ppalaga Here we go again! :D

- Fixes editorconfig/editorconfig#375, which has the discussion leading up to this PR.
- Replaces #22 per https://github.com/editorconfig/editorconfig-core-test/pull/22#issuecomment-435235532

- Adds new `new_ec_test_multiline` function that adds a test accepting multiple
  output lines in any order.  The matching regex for this function must be specified in CMake's sort order - see discussion at https://gitlab.kitware.com/cmake/cmake/issues/18551
- Rewrites the pertinent tests to use `new_ec_test_multiline`.
- Bumps the required CMake version up to 3.5.0 per @xuhdev at https://github.com/editorconfig/editorconfig-core-test/pull/22#discussion_r228741464

[editorconfig-core-c](https://github.com/editorconfig/editorconfig-core-c) and my [editorconfig-core-vimscript](https://github.com/cxw42/editorconfig-core-vimscript) both pass 100% of tests on Cygwin **(edit) and Windows** with this PR in place (except for utf-8 filenames).

Note: also includes one commit for the benefit of Cygwin (my development environment) to not run tests that can't possibly succeed on Cygwin.

**ppalaga**: Would you please test your Java CI?

**Edit** I have now run these tests on mingw-w64 x86 with Windows CMake 3.6.1 in cmd.  All pass except for utf_8_char, which I didn't touch :) .  I see per [this](https://ci.appveyor.com/project/ppalaga/editorconfig-core-c/build/1.0.92?fullLog=true#L1049) (linked from https://github.com/editorconfig/editorconfig-core-c/pull/31#issue-154810185) that this is a known issue on Windows.

xuhdev commented on filetree/CMakeLists.txt:
```diff
@@ -50,12 +52,13 @@ new_ec_test(root_file_mixed_case root_file.in root_mixed/test.a "^child=true[ \t
 
 # Windows style path separator in the command line should work on Windows, but
 # should not work on other systems
+# TODO figure out how to handle Cygwin
```
Can Cygwin automatically recognize backslashes as slashes?

ffes replied:
No, it can't. Cygwin is a POSIX-compatible layer, so it expects slashes. That is what `cygpath` is for. To convert Windows paths to Cygwin paths, including the right mountpoint.

The is a very similar question at the Cygwin mailing list at the moment: https://www.cygwin.com/ml/cygwin/2018-11/msg00141.html

cxw42 replied:
I am also not aware of a way.  CMake does not define WIN32 on Cygwin, so this conditional skips the test on Cygwin. 

@ffes Thanks for the link.  That makes me think that perhaps I should just remove the "TODO" line.

xuhdev replied:
Perhaps we should just skip the related tests :)

cxw42 replied:
@xuhdev We certainly could, and I am happy to make that edit.  However, the test is meaningful, to the extent that this is actually a difference in platform-specific behaviour.  I propose leaving the test, but removing the TODO line, and adding `cmake_minimum_required` so we are guaranteed that Cygwin doesn't set `WIN32` ([CMake commit reference](https://gitlab.kitware.com/cmake/cmake/commit/85c0a69a92e78275ea0b180482bafcdb877b0dc3)).  What do you think?

xuhdev replied:
I looked at it again. Yes, I agree with you that the test is meaningful. If cygwin does not specifically handle backslashes in paths, then in principle I think we should treat it like a UNIX system, i.e., backslash is backslash.

xuhdev replied:
Sure. Let's enforce 3.5

cxw42 replied:
@xuhdev Thanks!  I updated that, and rebased on the latest master.

cxw42 commented:
@ppalaga Have you had a chance to try out the Java core?  Unless this change causes problems for you, I think it's ready to merge.

cxw42 commented:
@xuhdev Fixed, plus I found and fixed a few `EQUAL` comparisons that should have been `STREQUAL`.  I re-ran {C core, Vimscript core} x {Cygwin, MinGW} and it looks good.  Thanks!

I updated the copyright dates across the board so no one else would have to do them for the rest of the year :) .

ppalaga commented:
There are 20 failing tests with ec4j:

* Windows: https://ci.appveyor.com/project/ppalaga/ec4j/builds/20452747?fullLog=true#L826 
* Linux: https://travis-ci.org/ec4j/ec4j/jobs/457809294#L3021

Let me check what is the problem.

ppalaga commented:
Here is the detailed output for the first failure `meta_multiline`:

```
2/189 Testing: meta_multiline
2/189 Test: meta_multiline
Command: "C:/Program Files (x86)/CMake/bin/cmake.exe" "-D" "EDITORCONFIG_CMD=C:/Program Files/Java/jdk1.7.0/bin/java.exe;-cp;C:/projects/ec4j/core/target/ec4j-core.jar;org.ec4j.core.cli.Cli" "-D" "ECARGS:LIST=-f;meta.in;C:/projects/ec4j/core/editorconfig-core-test/meta/meta.c" "-P" "C:/projects/ec4j/core/editorconfig-core-test/cmake/ec_sort.cmake"
Directory: C:/projects/ec4j/core/editorconfig-core-test/meta
"meta_multiline" start time: Nov 21 07:01 Coordinated Universal Time
Output:
----------------------------------------------------------
CMake Error at C:/projects/ec4j/core/editorconfig-core-test/cmake/ec_sort.cmake:55 (message):
  C:/Program
  Files/Java/jdk1.7.0/bin/java.exe;-cp;C:/projects/ec4j/core/target/ec4j-core.jar;org.ec4j.core.cli.Cli
  -f;meta.in;C:/projects/ec4j/core/editorconfig-core-test/meta/meta.c
  returned a nonzero exit code
<end of output>
```

source: https://ci.appveyor.com/project/ppalaga/ec4j/builds/20452747?fullLog=true#L894

Maybe the reason is that my `EDITORCONFIG_CMD` consists of `java` executable and some space separated arguments? Here is its definition: 
https://github.com/ec4j/ec4j/blob/8b8fa5578e293bf2e209432f2a1a399ff23fe36f/core/CMakeLists.txt#L20

Should I change something there?

ppalaga commented:
When I to the proper directory, replace semicolons with spaces in the command that `returned a nonzero exit code` above, and run it, it works properly. It returns `answer=42` as expected. 

I am not sure where the semicolons come from. Maybe they should not be there?

cxw42 commented:
@ppalaga Thanks!  Would you please try the new commit?  Before doing so, in your Travis/Appveyor config, would you please add `--output-on-failure` to the` ctest` command line?  That will give us a bit more information on failures.

The semicolons in `.../java.exe;-cp;C:/projects/...` are added by CMake.  This is because the definition

    set(EDITORCONFIG_CMD ${Java_JAVA_EXECUTABLE} -cp ${PROJECT_SOURCE_DIR}/target/ec4j-core.jar org.ec4j.core.cli.Cli)

creates a list-valued variable, and lists in CMake are just strings with semicolons between list items ([reference](https://cmake.org/cmake/help/v3.5/manual/cmake-language.7.html#lists)).  By contrast, `set(CMD "${Java} -cp ${...}")` (with `""`) would create a string-valued variable.

I have updated `cmake/ec_sort.cmake` to accept list-valued variables or single strings.  It should now pass the tests using your current `EDITORCONFIG_CMD`.  Please let me know.  Much appreciated!

ppalaga commented:
Yes, with eccf9d4 I see ec4j passing locally. Let's see if both CI jobs will pass too https://ci.appveyor.com/project/ppalaga/ec4j/builds/20462289 https://travis-ci.org/ec4j/ec4j/jobs/457960159

xuhdev commented:
Great work! Thanks!
